app/controllers/image_controller.rb

```diff
@@ -364,62 +410,96 @@ def process_image(args, upload)
-      init_project_vars_for_add_or_edit(@image)
+    set_licenses_ivar
+    check_image_permission
+    if request.method == "POST"
+      update_image
     else
-      @image.attributes = whitelisted_image_params
-      xargs = {}
-      xargs[:set_date] = @image.when if @image.when_changed?
-      xargs[:set_notes] = @image.notes if @image.notes_changed?
-      if @image.copyright_holder_changed?
-        xargs[:set_copyright_holder] = @image.copyright_holder
-      end
-      if @image.original_name_changed?
-        xargs[:set_original_name] = @image.original_name
-      end
-      xargs[:set_license] = @image.license if @image.license_id_changed?
-      done = false
-      if xargs.empty?
-        if update_projects(@image, params[:project])
-          flash_notice(:runtime_image_edit_success.t(id: @image.id))
-        else
-          flash_notice(:runtime_no_changes.t)
-        end
-        done = true
-      elsif !@image.save
-        flash_object_errors(@image)
-      else
-        xargs[:id] = @image
-        @image.log_update
-        flash_notice(:runtime_image_edit_success.t(id: @image.id))
-        update_projects(@image, params[:project])
-        done = true
-      end
-      if done
-        redirect_with_query(action: "show_image", id: @image.id)
-      else
-        init_project_vars_for_reload(@image)
-      end
+      init_project_vars_for_add_or_edit(@image)
     end
   end
 
   ##############################################################################
 
   private
 
+  def update_image
+    return unless image_exists
+
+    set_licenses_ivar
+    check_image_permission
+
+    @image.attributes = whitelisted_image_params
+    xargs = set_xargs
+
+    done = figure_out_if_done(xargs)
+    if done
+      redirect_with_query(action: "show_image", id: @image.id)
+    else
+      init_project_vars_for_reload(@image)
+    end
+  end
+
+  def set_licenses_ivar
+    @licenses = License.current_names_and_ids(@image.license)
+  end
+
+  def check_image_permission
+    redirect_with_query(action: "show_image", id: @image) unless
+      check_permission!(@image)
+  end
+
+  def set_xargs
+    xargs = {}
+    xargs[:set_date] = @image.when if @image.when_changed?
+    xargs[:set_notes] = @image.notes if @image.notes_changed?
+    if @image.copyright_holder_changed?
+      xargs[:set_copyright_holder] = @image.copyright_holder
+    end
+    if @image.original_name_changed?
+      xargs[:set_original_name] = @image.original_name
+    end
+    xargs[:set_license] = @image.license if @image.license_id_changed?
+    xargs
+  end
+
+  def figure_out_if_done(xargs = {})
+    if xargs.empty?
+      maybe_update_projects
+      true
+    elsif !@image.save
+      flash_object_errors(@image)
+      false
+    else
+      xargs[:id] = @image
+      @image.log_update
+      flash_notice(:runtime_image_edit_success.t(id: @image.id))
+      update_projects(@image, params[:project])
+      true
+    end
+  end
+
+  def maybe_update_projects
```

Comment:
Haha!  Great name!  ...If your goal is to make readers laugh.  Less great if you're interested in them actually understanding what this method does. :)

Reply:
Right, this always attempts to update projects, and renders an error if it fails, so "update_projects!" would be a whole lot more clear.

Reply:
`my_goal_is_always_to_make_readers_laugh!`

Reply:
"maybe_update_projects_maybe_not_whats_it_to_you_punk?"